File: lib/util/name.ml
(* Copyright (c) 1998-2007 INRIA *)

(* La liste des caractères interdits *)
let forbidden_char = [ ':'; '@'; '#'; '='; '$' ]

(* Name.lower *)

let unaccent_utf_8 lower s i =
  let fns =
    if lower then fun n s -> (String.lowercase_ascii s, n) else fun n s -> (s, n)
  in
  let fnc =
    if lower then fun n c -> (String.make 1 @@ Char.lowercase_ascii c, n)
    else fun n c -> (String.make 1 c, n)
  in
  let s, n =
    Unidecode.decode fns fnc
      (fun n -> (String.sub s i (n - i), n))
      s i (String.length s)
  in
  if lower then (String.lowercase_ascii s, n) else (s, n)

let next_chars_if_equiv s i t j =
  if i >= String.length s || j >= String.length t then None
  else
    let s1, i1 = unaccent_utf_8 true s i in
    let t1, j1 = unaccent_utf_8 true t j in
    if s1 = t1 then Some (i1, j1) else None

(* Name.lower:
    - uppercase -> lowercase
    - no accents
    - chars no letters and no numbers (except '.') => spaces (stripped)
     Key comparison (first name, surname, number) applies "lower" equality
     on first names and surnames *)
let lower s =
  let rec copy special i len =
    if i = String.length s then Buff.get len
    else if Char.code s.[i] < 0x80 then
      match s.[i] with
      | ('a' .. 'z' | 'A' .. 'Z' | '0' .. '9' | '.') as c ->
          let len = if special then Buff.store len ' ' else len in
          let c = Char.lowercase_ascii c in
          copy false (i + 1) (Buff.store len c)
      | _ -> copy (len <> 0) (i + 1) len
    else
      let len = if special then Buff.store len ' ' else len in
      let t, j = unaccent_utf_8 true s i in
      copy false j (Buff.mstore len t)
  in
  copy false 0 0

let title s =
  let t = ref true in
  let cmap u =
    let r = if !t then Uucp.Case.Map.to_upper u else Uucp.Case.Map.to_lower u in
    t := not (Uucp.Alpha.is_alphabetic u);
    r
  in
  Utf8.cmap_utf_8 cmap s

(* Name.abbrev *)

(* List of abbreviations. If abbreviation is mapped to [Some s] should be remplaced by
   [s]. If mapped to None, it should be removed from name. *)
let abbrev_list =
  [
    ("a", None);
    ("af", None);
    ("d", None);
    ("de", None);
    ("di", None);
    ("ier", Some "i");
    ("of", None);
    ("saint", Some "st");
    ("sainte", Some "ste");
    ("van", None);
    ("von", None);
    ("zu", None);
    ("zur", None);
  ]

(* Checks if the word starting at [i] in [s] is [p]. *)
let is_word s i p =
  let rec is_word s i p ip =
    if ip = String.length p then
      if i = String.length s then true else if s.[i] = ' ' then true else false
    else if i = String.length s then false
    else if s.[i] = p.[ip] then is_word s (i + 1) p (ip + 1)
    else false
  in
  is_word s i p 0

(* Checks if word that starts at position [i] in [s] is one of abbreviation *)
let rec search_abbrev s i = function
  | (w, a) :: pl ->
      if is_word s i w then Some (String.length w, a) else search_abbrev s i pl
  | [] -> None

(* Name.abbrev: suppress lowercase particles, shorten "saint" into "st" *)
let abbrev s =
  let rec copy can_start_abbrev i len =
    if i >= String.length s then Buff.get len
    else
      match s.[i] with
      | ' ' -> copy true (i + 1) (Buff.store len ' ')
      | c ->
          if can_start_abbrev then
            match search_abbrev s i abbrev_list with
            | None -> copy false (i + 1) (Buff.store len c)
            | Some (n, Some a) -> copy false (i + n) (Buff.mstore len a)
            | Some (n, None) -> copy true (i + n + 1) len
          else copy false (i + 1) (Buff.store len c)
  in
  copy true 0 0

(* Name.strip *)

(* Name.strip_c = name without the charater c given as parameter *)
let strip_c s c =
  let rec copy i len =
    if i = String.length s then Buff.get len
    else if s.[i] = c then copy (i + 1) len
    else copy (i + 1) (Buff.store len s.[i])
  in
  copy 0 0

let strip s = strip_c s ' '

(* String without any forbidden caracters defined in forbidden_char *)
(* ******************************************************************** *)
(*  [Fonc] purge : string -> string                                     *)

(* ******************************************************************** *)

(** [Description] : Supprime tous les caractères interdits (défini par
    forbidden_char) présents dans la chaine passée en argument. [Args] :
    - s : string que l'on veut purger [Retour] :
    - string : retourne la chaîne délestée des caractères interdits [Rem] :
      Exporté en clair hors de ce module. *)
let purge s = List.fold_left strip_c s forbidden_char

(* Name.crush *)

(* If string starting from [i] contains roman number then returns the next position,
   else returns None. *)
let roman_number s i =
  let rec loop i =
    if i = String.length s then Some i
    else if s.[i] = ' ' then Some i
    else match s.[i] with 'i' | 'v' | 'x' | 'l' -> loop (i + 1) | _ -> None
  in
  if i = 0 || s.[i - 1] = ' ' then loop i else None

(* Name.crush, a custom sonnex/soundex-like phonetic algorithm:
     - no spaces
     - roman numbers are keeped
     - vowels are suppressed, except in words starting with a vowel,
       where this vowel is converted into "e"
     - "k" and "q" replaced by "c"
     - "y" replaced by "i"
     - "z" replaced by "s"
     - "ph" replaced by "f"
     - others "h" deleted
     - s at end of words are deleted
     - no double lowercase consons *)
let crush s =
  let rec copy i len first_vowel =
    if i = String.length s then Buff.get len
    else if s.[i] = ' ' then copy (i + 1) len true
    else
      match roman_number s i with
      | Some j ->
          let rec loop i len =
            if i = j then copy j len true
            else loop (i + 1) (Buff.store len s.[i])
          in
          loop i len
      | _ -> (
          match s.[i] with
          | 'a' | 'e' | 'i' | 'o' | 'u' | 'y' ->
              let len = if first_vowel then Buff.store len 'e' else len in
              copy (i + 1) len false
          | 'h' ->
              let len =
                if i > 0 && s.[i - 1] = 'p' then Buff.store (len - 1) 'f'
                else len
              in
              copy (i + 1) len first_vowel
          | ('s' | 'z') when i = String.length s - 1 || s.[i + 1] = ' ' ->
              let len =
                let rec loop i len =
                  if
                    i > 0 && len > 0
                    && s.[i] = Bytes.get !Buff.buff len
                    && (s.[i] = 's' || s.[i] = 'z')
                  then loop (i - 1) (len - 1)
                  else len + 1
                in
                loop (i - 1) (len - 1)
              in
              copy (i + 1) len false
          | 's' when i = String.length s - 1 || s.[i + 1] = ' ' ->
              copy (i + 1) len false
          | c ->
              if i > 0 && s.[i - 1] = c then copy (i + 1) len false
              else
                let c = match c with 'k' | 'q' -> 'c' | 'z' -> 's' | c -> c in
                copy (i + 1) (Buff.store len c) false)
  in
  copy 0 0 true

(* strip_lower *)

(* strip_lower = strip o lower, as first comparison of names.
   First names and Surnames comparison is strip_lower equality. *)
let strip_lower s = strip (lower s)

(* crush_lower *)

(* crush_lower = crush o abbrev o lower, as second comparison of names.
   In index by names, the "names" are crush_lowers *)
let crush_lower s = crush (abbrev (lower s))

(* concat two strings using Bytes module *)
let concat_aux fn l1 sn l2 =
  let b = Bytes.create (l1 + l2 + 1) in
  Bytes.blit_string fn 0 b 0 l1;
  Bytes.blit_string sn 0 b (l1 + 1) l2;
  Bytes.unsafe_set b l1 ' ';
  Bytes.unsafe_to_string b

let concat fn sn = concat_aux fn (String.length fn) sn (String.length sn)
let contains_forbidden_char s = List.exists (String.contains s) forbidden_char

(* Copy/paste from String.split_on_char adapted to our needs *)
let split_sname_callback fn s =
  let open String in
  let j = ref (length s) in
  for i = length s - 1 downto 0 do
    if match unsafe_get s i with ' ' | '-' -> true | _ -> false then (
      fn (i + 1) (!j - i - 1);
      j := i)
  done;
  fn 0 !j

(* Copy/paste from String.split_on_char adapted to our needs *)
let split_fname_callback fn s =
  let open String in
  let j = ref (length s) in
  for i = length s - 1 downto 0 do
    if unsafe_get s i = ' ' then (
      fn (i + 1) (!j - i - 1);
      j := i)
  done;
  fn 0 !j

let split_sname s =
  let r = ref [] in
  split_sname_callback (fun i j -> r := String.sub s i j :: !r) s;
  !r

let split_fname s =
  let r = ref [] in
  split_fname_callback (fun i j -> r := String.sub s i j :: !r) s;
  !r

File: src/python/lib/util/name.py
"""
Name processing utilities converted from OCaml.
Handles name normalization, abbreviations, and phonetic matching.
"""

import re
import unicodedata
from typing import Callable, List, Optional, Tuple


class Buff:
    """Buffer utility class for efficient string building."""

    def __init__(self):
        self.buffer = []

    def store(self, char: str) -> None:
        """Store a character in the buffer."""
        self.buffer.append(char)

    def mstore(self, text: str) -> None:
        """Store multiple characters in the buffer."""
        self.buffer.extend(text)

    def get(self) -> str:
        """Get the final string from buffer."""
        return "".join(self.buffer)

    def length(self) -> int:
        """Get current buffer length."""
        return len(self.buffer)


# List of forbidden characters
FORBIDDEN_CHAR = [":", "@", "#", "=", "$"]

# List of abbreviations. If abbreviation is mapped to a string, should be replaced by
# that string. If mapped to None, it should be removed from name.
ABBREV_LIST = [
    ("a", None),
    ("af", None),
    ("d", None),
    ("de", None),
    ("di", None),
    ("ier", "i"),
    ("of", None),
    ("saint", "st"),
    ("sainte", "ste"),
    ("van", None),
    ("von", None),
    ("zu", None),
    ("zur", None),
]


def unaccent_utf_8(lower_case: bool, s: str, i: int) -> Tuple[str, int]:
    """
    Remove accents from UTF-8 string starting at position i.
    Returns the processed character(s) and the next position.
    """
    if i >= len(s):
        return "", i

    char = s[i]
    # Handle ASCII characters
    if ord(char) < 0x80:
        result = char.lower() if lower_case else char
        return result, i + 1

    # Handle Unicode characters - remove accents
    try:
        # Normalize to NFD (decomposed form) and remove combining characters
        normalized = unicodedata.normalize("NFD", char)
        ascii_char = "".join(c for c in normalized if unicodedata.category(c) != "Mn")
        if not ascii_char:
            ascii_char = char
        result = ascii_char.lower() if lower_case else ascii_char
        return result, i + 1
    except:
        result = char.lower() if lower_case else char
        return result, i + 1


def next_chars_if_equiv(s: str, i: int, t: str, j: int) -> Optional[Tuple[int, int]]:
    """
    Check if characters at positions i and j are equivalent after unaccenting.
    Returns next positions if equivalent, None otherwise.
    """
    if i >= len(s) or j >= len(t):
        return None

    s1, i1 = unaccent_utf_8(True, s, i)
    t1, j1 = unaccent_utf_8(True, t, j)

    if s1 == t1:
        return (i1, j1)
    return None


def lower(s: str) -> str:
    """
    Name.lower:
    - uppercase -> lowercase
    - no accents
    - chars no letters and no numbers (except '.') => spaces (stripped)
    Key comparison (first name, surname, number) applies "lower" equality
    on first names and surnames
    """
    buff = Buff()
    special = False
    i = 0

    while i < len(s):
        if ord(s[i]) < 0x80:
            char = s[i]
            if char.isalnum() or char == ".":
                if special:
                    buff.store(" ")
                buff.store(char.lower())
                special = False
            else:
                special = buff.length() != 0
            i += 1
        else:
            if special:
                buff.store(" ")
            text, j = unaccent_utf_8(True, s, i)
            buff.mstore(text)
            special = False
            i = j

    return buff.get()


def title(s: str) -> str:
    """
    Convert string to title case, handling Unicode properly.
    """
    result = []
    capitalize_next = True

    for char in s:
        if char.isalpha():
            if capitalize_next:
                result.append(char.upper())
                capitalize_next = False
            else:
                result.append(char.lower())
        else:
            result.append(char)
            capitalize_next = True

    return "".join(result)


def is_word(s: str, i: int, p: str) -> bool:
    """
    Checks if the word starting at [i] in [s] is [p].
    """
    if i + len(p) > len(s):
        return False

    # Check if the substring matches
    if s[i : i + len(p)] != p:
        return False

    # Check word boundary
    end_pos = i + len(p)
    if end_pos == len(s):
        return True
    elif s[end_pos] == " ":
        return True
    else:
        return False


def search_abbrev(
    s: str, i: int, abbrev_list: List[Tuple[str, Optional[str]]]
) -> Optional[Tuple[int, Optional[str]]]:
    """
    Checks if word that starts at position [i] in [s] is one of abbreviation.
    """
    for word, abbrev in abbrev_list:
        if is_word(s, i, word):
            return (len(word), abbrev)
    return None


def abbrev(s: str) -> str:
    """
    Name.abbrev: suppress lowercase particles, shorten "saint" into "st"
    """
    buff = Buff()
    can_start_abbrev = True
    i = 0

    while i < len(s):
        if s[i] == " ":
            buff.store(" ")
            can_start_abbrev = True
            i += 1
        else:
            if can_start_abbrev:
                result = search_abbrev(s, i, ABBREV_LIST)
                if result is None:
                    buff.store(s[i])
                    can_start_abbrev = False
                    i += 1
                else:
                    n, replacement = result
                    if replacement is not None:
                        buff.mstore(replacement)
                        can_start_abbrev = False
                        i += n
                    else:
                        # Skip the abbreviation and the following space
                        can_start_abbrev = True
                        i += n
                        if i < len(s) and s[i] == " ":
                            i += 1
            else:
                buff.store(s[i])
                can_start_abbrev = False
                i += 1

    return buff.get()


def strip_c(s: str, c: str) -> str:
    """
    Name.strip_c = name without the character c given as parameter
    """
    return s.replace(c, "")


def strip(s: str) -> str:
    """
    Remove all spaces from string.
    """
    return strip_c(s, " ")


def purge(s: str) -> str:
    """
    String without any forbidden characters defined in forbidden_char.
    Removes all forbidden characters from the string.
    """
    result = s
    for char in FORBIDDEN_CHAR:
        result = strip_c(result, char)
    return result


def roman_number(s: str, i: int) -> Optional[int]:
    """
    If string starting from [i] contains roman number then returns the next position,
    else returns None.
    """
    if i == 0 or (i > 0 and s[i - 1] == " "):
        j = i
        while j < len(s):
            if s[j] == " ":
                return j
            elif s[j] in "ivxl":
                j += 1
            else:
                return None
        return j
    return None


def crush(s: str) -> str:
    """
    Name.crush, a custom sonnex/soundex-like phonetic algorithm:
    - no spaces
    - roman numbers are kept
    - vowels are suppressed, except in words starting with a vowel,
      where this vowel is converted into "e"
    - "k" and "q" replaced by "c"
    - "y" replaced by "i"
    - "z" replaced by "s"
    - "ph" replaced by "f"
    - others "h" deleted
    - s at end of words are deleted
    - no double lowercase consonants
    """
    buff = Buff()
    first_vowel = True
    i = 0

    while i < len(s):
        if s[i] == " ":
            first_vowel = True
            i += 1
        else:
            roman_end = roman_number(s, i)
            if roman_end is not None:
                # Copy roman number as-is
                while i < roman_end:
                    buff.store(s[i])
                    i += 1
                first_vowel = True
            else:
                char = s[i]
                if char in "aeiouy":
                    if first_vowel:
                        buff.store("e")
                    first_vowel = False
                    i += 1
                elif char == "h":
                    # Handle "ph" -> "f"
                    if i > 0 and s[i - 1] == "p" and buff.length() > 0:
                        # Remove the 'p' and add 'f'
                        buff.buffer = buff.buffer[:-1]
                        buff.store("f")
                    # Otherwise skip 'h'
                    i += 1
                elif char in "sz" and (
                    i == len(s) - 1 or (i + 1 < len(s) and s[i + 1] == " ")
                ):
                    # Remove trailing 's' or 'z' at end of words
                    # Also remove duplicate 's' or 'z'
                    while (
                        buff.length() > 0
                        and buff.buffer[-1] in "sz"
                        and buff.buffer[-1] == char
                    ):
                        buff.buffer = buff.buffer[:-1]
                    i += 1
                elif char == "s" and (
                    i == len(s) - 1 or (i + 1 < len(s) and s[i + 1] == " ")
                ):
                    # Skip trailing 's'
                    i += 1
                else:
                    # Handle character replacements and avoid duplicates
                    if i > 0 and s[i - 1] == char:
                        # Skip duplicate
                        pass
                    else:
                        # Apply character replacements
                        if char == "k" or char == "q":
                            char = "c"
                        elif char == "z":
                            char = "s"
                        buff.store(char)
                    first_vowel = False
                    i += 1

    return buff.get()


def strip_lower(s: str) -> str:
    """
    strip_lower = strip o lower, as first comparison of names.
    First names and Surnames comparison is strip_lower equality.
    """
    return strip(lower(s))


def crush_lower(s: str) -> str:
    """
    crush_lower = crush o abbrev o lower, as second comparison of names.
    In index by names, the "names" are crush_lowers
    """
    return crush(abbrev(lower(s)))


def concat(fn: str, sn: str) -> str:
    """
    Concatenate two strings with a space.
    """
    return f"{fn} {sn}"


def contains_forbidden_char(s: str) -> bool:
    """
    Check if string contains any forbidden character.
    """
    return any(char in s for char in FORBIDDEN_CHAR)


def split_sname_callback(fn: Callable[[int, int], None], s: str) -> None:
    """
    Split surname on spaces and hyphens, calling callback for each part.
    """
    j = len(s)
    for i in range(len(s) - 1, -1, -1):
        if s[i] in " -":
            fn(i + 1, j - i - 1)
            j = i
    fn(0, j)


def split_fname_callback(fn: Callable[[int, int], None], s: str) -> None:
    """
    Split first name on spaces, calling callback for each part.
    """
    j = len(s)
    for i in range(len(s) - 1, -1, -1):
        if s[i] == " ":
            fn(i + 1, j - i - 1)
            j = i
    fn(0, j)


def split_sname(s: str) -> List[str]:
    """
    Split surname into parts separated by spaces or hyphens.
    """
    result = []
    split_sname_callback(lambda i, length: result.append(s[i : i + length]), s)
    return list(reversed(result))


def split_fname(s: str) -> List[str]:
    """
    Split first name into parts separated by spaces.
    """
    result = []
    split_fname_callback(lambda i, length: result.append(s[i : i + length]), s)
    return list(reversed(result))


# Example usage and tests
if __name__ == "__main__":
    # Test the functions
    test_name = "Jean-Baptiste de Saint-Exupéry"

    print(f"Original: {test_name}")
    print(f"Lower: {lower(test_name)}")
    print(f"Abbrev: {abbrev(lower(test_name))}")
    print(f"Crush: {crush(abbrev(lower(test_name)))}")
    print(f"Strip lower: {strip_lower(test_name)}")
    print(f"Crush lower: {crush_lower(test_name)}")

    print(f"Split surname: {split_sname('Martin-Dupont')}")
    print(f"Split first name: {split_fname('Jean Baptiste')}")

    print(f"Contains forbidden char: {contains_forbidden_char('test@example.com')}")
    print(f"Purged: {purge('test@example.com')}")
